File: tests/R/test_analysis.R
# Invariants of the R analysis outputs (output/*_r.csv).
suppressPackageStartupMessages({
  library(testthat)
  library(readr)
})

out <- file.path(root, "output")

test_that("WS value is zero-sum across the 30 franchises", {
  teams <- read_csv(file.path(out, "teams_r.csv"), show_col_types = FALSE)
  expect_lt(abs(sum(teams$value_ws)), 5)
  expect_equal(nrow(teams), 30)
})

test_that("every pick curve declines monotonically", {
  curve <- read_csv(file.path(out, "pick_curve_r.csv"), show_col_types = FALSE)
  expect_equal(nrow(curve), 60)
  for (col in c("exp_ws", "exp_min", "exp_vorp", "exp_peak3")) {
    expect_true(all(diff(curve[[col]]) <= 0), info = col)
  }
})

test_that("confidence intervals bracket the estimate", {
  teams <- read_csv(file.path(out, "teams_r.csv"), show_col_types = FALSE)
  expect_true(all(teams$ci_lo <= teams$value_ws & teams$value_ws <= teams$ci_hi))
})

test_that("kept minutes are a valid share", {
  teams <- read_csv(file.path(out, "teams_r.csv"), show_col_types = FALSE)
  expect_true(all(teams$kept_share >= 0 & teams$kept_share <= 1))
  expect_true(all(teams$kept_min >= 0))
})

test_that("college hit rates are consistent", {
  colleges <- read_csv(file.path(out, "colleges_r.csv"), show_col_types = FALSE)
  expect_true(all(colleges$hits >= 0 & colleges$hits <= colleges$draftees))
  expect_true(all(colleges$draftees >= 8))
  expect_equal(colleges$hit_rate, round(colleges$hits / colleges$draftees, 3))
})

test_that("steals are positive and ranked", {
  steals <- read_csv(file.path(out, "steals_r.csv"), show_col_types = FALSE)
  expect_equal(nrow(steals), 15)
  expect_true(all(steals$value_ws > 0))
  expect_equal(steals$value_ws, sort(steals$value_ws, decreasing = TRUE))
})

File: tests/python/test_analysis.py
"""Invariants of the Python analysis outputs (output/*.csv)."""
import csv
from pathlib import Path

OUT = Path(__file__).resolve().parents[2] / "output"


def rows(name):
    with (OUT / f"{name}.csv").open() as f:
        return list(csv.DictReader(f))


def test_ws_value_is_zero_sum_across_teams():
    # The isotonic fit preserves the weighted mean, so value above slot
    # summed over all 30 franchises must cancel to ~0 (rounding aside).
    total = sum(float(t["value_ws"]) for t in rows("teams"))
    assert abs(total) < 5.0
    assert len(rows("teams")) == 30


def test_pick_curves_decline_monotonically():
    curve = rows("pick_curve")
    assert len(curve) == 60
    for col in ("exp_ws", "exp_min", "exp_vorp", "exp_peak3"):
        means = [float(r[col]) for r in curve]
        assert all(a >= b for a, b in zip(means, means[1:])), col


def test_confidence_intervals_bracket_the_estimate():
    for t in rows("teams"):
        assert float(t["ci_lo"]) <= float(t["value_ws"]) <= float(t["ci_hi"])


def test_kept_minutes_are_a_valid_share():
    for t in rows("teams"):
        assert 0.0 <= float(t["kept_share"]) <= 1.0
        assert float(t["kept_min"]) >= 0.0


def test_college_hit_rates_are_consistent():
    for c in rows("colleges"):
        draftees, hits = int(c["draftees"]), int(c["hits"])
        assert 0 <= hits <= draftees
        assert draftees >= 8
        assert abs(float(c["hit_rate"]) - round(hits / draftees, 3)) < 1e-9


def test_steals_are_positive_and_ranked():
    values = [float(s["value_ws"]) for s in rows("steals")]
    assert len(values) == 15
    assert all(v > 0 for v in values)
    assert values == sorted(values, reverse=True)
